[aot_inductor] replace TORCH_CHECK with AOTI_CHECK in the generate cpp code

chenyang78 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119220

In some cases where we have TORCH_CHECK in loops, it may cause the host
compiler to spend hours optimizing the run_impl function. This PR
mitigated the issue by replacing TORCH_CHECK with a custom AOTI_CHECK,
where we force the underneath assert function to be noinline.

If forcing noinline caused any serious perf regression, we could
either add an option to turn on/off enable noinline. Or, we could
another an option to just turn AOTI_CHECK into a no-op, similar
to the ```assert``` macro from cassert.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @kadeng @muchulee8 @aakhundov @ColinPeppler

Commit:
[aot_inductor] replace TORCH_CHECK with AOTI_CHECK in the generate cpp code

In some cases where we have TORCH_CHECK in loops, it may cause the host
compiler to spend hours optimizing the run_impl function. This PR
mitigated the issue by replacing TORCH_CHECK with a custom AOTI_CHECK,
where we force the underneath assert function to be noinline.

If forcing noinline caused any serious perf regression, we could
either add an option to turn on/off enable noinline. Or, we could
another an option to just turn AOTI_CHECK into a no-op, similar
to the ```assert``` macro from cassert.

[ghstack-poisoned]

chenyang78 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1730,7 +1730,10 @@ def codegen_loops(self, code, worksharing):
     def assert_function(self) -> str:
-        return "TORCH_CHECK"
+        if V.graph.aot_mode:
+            return "AOTI_CHECK"
```
I think the change should be fine for both cpu and gpu.

hl475 reviewed (approved):
Thanks Yang!

desertfire commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1730,7 +1730,10 @@ def codegen_loops(self, code, worksharing):
+        if V.graph.aot_mode:
+            return "AOTI_CHECK"
+        else:
+            return "TORCH_CHECK"
```
JIT Inductor will also face the compilation time challenge sooner or later, so it's worth to unify the behavior here.

chenyang78 replied:
So, we can just remove the ```if V.graph.aot_mode``` condition? Thanks.

desertfire commented on torch/csrc/inductor/aoti_runtime/model.h:
```diff
@@ -15,6 +15,7 @@
 // in model.so, and should not refer to any aten/c10 headers except the stable
 // C ABI defined in torch/csrc/inductor/aoti_torch/c/shim.h. The same rule
 // applies to other files under torch/csrc/inductor/aoti_runtime/.
+#include <c10/util/Exception.h>
```
There is `c10/util/Exception.cpp`, so this has a potential to break the ABI-compatibility. For AOTI, we can implement a separate version of `torchCheckFail`. w.r.t. my comments earlier, we can probably start with fixing AOT first, and then JIT.

chenyang78 replied:
Hmm, I thought ```c10/util/Exception.h``` was indirectly included via ```c10/utils/generic_math.h``` and this was why I included the same header here. In any case, let me have our own version of torchCheckFail. Thanks.

Commit:
Update on "[aot_inductor] replace TORCH_CHECK with AOTI_CHECK in the generate cpp code"


In some cases where we have TORCH_CHECK in loops, it may cause the host
compiler to spend hours optimizing the run_impl function. This PR
mitigated the issue by replacing TORCH_CHECK with a custom AOTI_CHECK,
where we force the underneath assert function to be noinline.

If forcing noinline caused any serious perf regression, we could
either add an option to turn on/off enable noinline. Or, we could
another an option to just turn AOTI_CHECK into a no-op, similar
to the ```assert``` macro from cassert.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_runtime/model.h:
```diff
@@ -77,6 +78,42 @@ AOTI_NOINLINE static void throw_exception(
+  }
+}
+
+#ifdef STRIP_ERROR_MESSAGES
```
We need to update codecache.py to take this macro, otherwise users might be surprised by this behavior.

chenyang78 replied:
> We need to update codecache.py to take this macro, otherwise users might be surprised.

Could you elaborate? Do you mean we need to pass this macro from the compiler invocation command-line? I think this macro is only used for production mobile builds.

desertfire replied:
I think we can implement by implementing `__aoti_check` and `AOTI_CHECK` in C shim, and then naturally that will take whatever macro was used for building PyTorch. Another benefit of that is we can still call the same implementation in c10 (except we wrap with AOTI_NOINLINE) instead of having a potentially diverged behavior.

Commit:
Update on "[aot_inductor] replace TORCH_CHECK with AOTI_CHECK in the generate cpp code"


In some cases where we have TORCH_CHECK in loops, it may cause the host
compiler to spend hours optimizing the run_impl function. This PR
mitigated the issue by replacing TORCH_CHECK with a custom AOTI_CHECK,
where we force the underneath assert function to be noinline.

If forcing noinline caused any serious perf regression, we could
either add an option to turn on/off enable noinline. Or, we could
another an option to just turn AOTI_CHECK into a no-op, similar
to the ```assert``` macro from cassert.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "[aot_inductor] replace TORCH_CHECK with AOTI_CHECK in the generate cpp code"


In some cases where we have TORCH_CHECK in loops, it may cause the host
compiler to spend hours optimizing the run_impl function. This PR
mitigated the issue by replacing TORCH_CHECK with a custom AOTI_CHECK,
where we force the underneath assert function to be noinline.

If forcing noinline caused any serious perf regression, we could
either add an option to turn on/off enable noinline. Or, we could
another an option to just turn AOTI_CHECK into a no-op, similar
to the ```assert``` macro from cassert.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -487,6 +487,31 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_proxy_executor_call_function(
+      __func__,                        \
+      __FILE__,                        \
+      static_cast<uint32_t>(__LINE__), \
+      TORCH_CHECK_MSG(cond, "", __VA_ARGS__));
```
Wait, there is still `TORCH_CHECK_MSG`. Hmm, this is annoying, and seems like we have to pass var args. I wonder if it will be easier to just make `c10/util/Exception.h` header only (merging `c10/util/Exception.cpp` with `c10/util/Exception.h`). 

@ezyang , do you have any suggestion for this?

chenyang78 replied:
By reading c10's utility functions, seems this TORCH_CHECK_MSG is header only. After it's expanded, it invokes ```torchCheckMsgImpl```:

https://github.com/pytorch/pytorch/blob/56364124af8fe148ba8b0c935571ebae6500f33b/c10/util/Exception.h#L451-L463

where ```::c10::str``` also seems to be header only:

https://github.com/pytorch/pytorch/blob/56364124af8fe148ba8b0c935571ebae6500f33b/c10/util/StringUtil.h#L114-L118

desertfire replied:
I see. Then it should be ok for this PR. At some point, we need to come up with a small set of self-contained `c10/util` headers.

chenyang78 replied:
> I see. Then it should be ok for this PR. At some point, we need to come up with a small set of self-contained `c10/util` headers.

Yes. That would be very helpful.

desertfire reviewed (changes requested):
`TORCH_CHECK_MSG` still needs to be addressed

ColinPeppler commented:
hello @chenyang78, n00b q: how to figure out why the host compiler is spending a lot of time optimizing due to inlining?

chenyang78 commented:
> hello @chenyang78, n00b q: how to figure out why the host compiler is spending a lot of time optimizing due to inlining?

I guess it mostly comes from my intuition :), because inlining increases the size of the function body, which could result in longer compilation time.
